### PR TITLE
LCOW: WORKDIR correct handling

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard <jhoward@microsoft.com>

WORKDIR in LCOW mode was being treated as a Windows path and not working - `/bin` was being translated to `C:\bin`. This fixes it to use Unix semantics.

```PowerShell
PS E:\docker\build\lcow> docker build -f workdir --no-cache .
Sending build context to Docker daemon  6.144kB
Step 1/2 : FROM busybox
 ---> efe10ee6727f
Step 2/2 : WORKDIR /bin
 ---> 29731c1d2cbd
Removing intermediate container 7425c28fb858
Successfully built 29731c1d2cbd
PS E:\docker\build\lcow> docker inspect -f "{{.ContainerConfig.Cmd}} -- {{.Config.WorkingDir}}" 297
[/bin/sh -c #(nop) WORKDIR /bin] -- /bin
PS E:\docker\build\lcow> docker run --rm 297 pwd
/bin
PS E:\docker\build\lcow> docker run --rm 297 ls -l ./sh
-rwxr-xr-x  385 root     root       1049592 Aug  4 18:44 ./sh
PS E:\docker\build\lcow>
```

@johnstep @dnephin PTAL

@Gupta-ak - I left a `TODO` in there for the remote filesystem operation.

### CODE REVIEW

File: builder/dockerfile/dispatchers_windows_test.go
Comment (johnstep) [inline builder/dockerfile/dispatchers_windows_test.go]: Since `normaliseWorkdirUnix` replaces Windows path separators with Unix, how about tests for that?

**Comment by jhowardmsft** [comment]
Updated for `docker run --workdir` to work as well.

```PowerShell
PS E:\docker\build\lcow> docker run --rm --workdir /bin busybox ls -l sh
-rwxr-xr-x  385 root     root       1049592 Aug  4 19:56 sh
```

File: builder/dockerfile/dispatchers_windows_test.go
Comment (jhowardmsft) [inline builder/dockerfile/dispatchers_windows_test.go]: Added case `{"linux", `\a`, `b\c`, `/a/b/c`, ``},`

File: daemon/container.go
Comment (johnstep) [inline daemon/container.go]: Is this a sufficient LCOW check? I think so, at least for now, but I recall usually seeing something more explicit, like `windows` and `system.LCOWSupported`.

Comment (jhowardmsft) [inline daemon/container.go]: I could be more explicit, but in reality, here it's redundant. It can only be LCOW.

**Review by johnstep — APPROVED** [review]
LGTM

**Comment by tonistiigi** [comment]
As discussed in https://github.com/moby/moby/issues/33854 there is no such thing as "build platform". Only platform for current stage (and maybe default platform for image access if maintainers can agree on it). So these should be used instead for detecting if conversion is needed. Ideally I would even like if this conversion only happens on starting a container instead of the builder we can discuss that in another issue. OCI image spec doesn't seem to define anything about `C:` prefix.

**Comment by jhowardmsft** [comment]
> As discussed in #33854 there is no such thing as "build platform".

To be more accurate, your goal is that there shouldn't be such a thing as a "build platform". I don't necessarily disagree (entirely), but the fact is there is presently, and the resolution of how the code base gets updated, if indeed to use `FROM --platform` is orthogonal to making progress on LCOW in the limited time available to hit RTM. This PR doesn't do anything contentious that isn't already in the codebase. 

> So these should be used instead for detecting if conversion is needed.

Yes, but after #33854 has been ratified and agreed.

> Ideally I would even like if this conversion only happens on starting a container instead of the builder but we can discuss that in another issue.

Sure, that's for later and also orthogonal. But I have concerns about cache breaks if you are going to consider that. The builder has done normalization in this way since WS2016 TP4 days - a very long time. This PR doesn't change that, it just makes sure that the Windows daemon doesn't universally assume it's a full Windows path.

> CI image spec doesn't seem to define anything about C: prefix.

Again, if it doesn't, that is also orthogonal to this PR and something which needs addressing in the OCI docs rather than here.

File: builder/dockerfile/dispatchers_unix_test.go
Comment (stevvooe) [inline builder/dockerfile/dispatchers_unix_test.go]: nit: `normalize` (do we standardize on us eng?)

File: builder/dockerfile/dispatchers_windows_test.go
Comment (stevvooe) [inline builder/dockerfile/dispatchers_windows_test.go]: "Normalize"

**Comment by tonistiigi** [comment]
As discussed offline, both build and create paths will need to be updated based on the #33854 outcome and the current definitely isn't correct. LGTM to unblock

**Comment by stevvooe** [comment]
@johnstep Are we going to have a PR correcting the spelling errors before merging?

**Comment by johnstep** [comment]
@stevvooe I also prefer we standardize on language _and_ locale. Is there official guidance for this somewhere? Can it be overridden on a file-by-file basis? @jhowardmsft, what is your take?

**Comment by stevvooe** [comment]
I took care of it in https://github.com/moby/moby/pull/34602.

**Pull request merged**
